src/backend/dx11/shaders/copy.hlsl

```diff
@@ -1,15 +1,200 @@
-cbuffer BufferImageCopy : register(b0) {
-    uint2 BufferSize;
-    uint2 ImageOffset;
+struct BufferCopy {
+    uint4 SrcDst;
 };
 
-StructuredBuffer<uint> CopySrc : register(t0);
-RWTexture2D<uint> CopyDst : register(u0);
+struct ImageCopy {
+    uint4 Src;
+    uint4 Dst;
+};
+
+struct BufferImageCopy {
+    // x=offset, yz=size
+    uint4 BufferVars;
+    uint4 ImageOffset;
+    uint4 ImageExtent;
+};
+
+cbuffer CopyConstants : register(b0) {
+    BufferCopy BufferCopies;
+    ImageCopy ImageCopies;
+    BufferImageCopy BufferImageCopies;
+};
+
+uint2 GetImageDst(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.ImageOffset.xy + dispatch_thread_id.xy;
+}
+
+uint2 GetImageSrc(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.ImageOffset.xy + dispatch_thread_id.xy;
+}
+
+uint GetBufferDst(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.BufferVars.x + dispatch_thread_id.x + dispatch_thread_id.y * BufferImageCopies.BufferVars.y;
+}
+
+uint GetBufferSrc(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.BufferVars.x + dispatch_thread_id.x + dispatch_thread_id.y * BufferImageCopies.BufferVars.y;
+}
+
+uint Uint4ToUint(uint4 data)
+{
+    data.x = min(data.x, 0x000000ff);
```

Comment:
nit: could probably vectorize this better, e.g.
```hlsl
return dot(min(data, 0xFF), 1 << uint4(0, 8, 16, 24));
```

Reply:
Beautiful! 😄